shadow-4.0.3/man/pwck.8
.\"$Id: pwck.8 6 2005-03-20 15:34:28Z bubulle $
.\" Copyright 1992, Julianne Frances Haugh
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of Julianne F. Haugh nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY JULIE HAUGH AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL JULIE HAUGH OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.TH PWCK 8
.SH NAME
pwck \- verify integrity of password files
.SH SYNOPSIS
\fBpwck\fR [\fB-sr\fR] [\fIpasswd\fR \fIshadow\fR]
.SH DESCRIPTION
\fBpwck\fR verifies the integrity of the system authentication information.
All entries in the \fI/etc/passwd\fR and \fI/etc/shadow\fR are checked to
see that the entry has the proper format and valid data in each field.
The user is prompted to delete entries that are improperly formatted or
which have other uncorrectable errors.
.P
Checks are made to verify that each entry has
.sp
.in +.5i
- the correct number of fields
.br
- a unique user name
.br
- a valid user and group identifier
.br
- a valid primary group
.br
- a valid home directory
.br
- a valid login shell
.in -.5i
.sp
.P
The checks for correct number of fields and unique user name are fatal.
If the entry has the wrong number of fields, the user will be prompted to
delete the entire line.
If the user does not answer affirmatively, all further checks are bypassed.
An entry with a duplicated user name is prompted for deletion, but the
remaining checks will still be made.
All other errors are warning and the user is encouraged to run the
\fBusermod\fR command to correct the error.
.P
The commands which operate on the \fI/etc/passwd\fR file are not able to
alter corrupted or duplicated entries.
\fBpwck\fR should be used in those circumstances to remove the offending
entry.
.SH OPTIONS
By default, \fBpwck\fR operates on the files \fI/etc/passwd\fR and
\fI/etc/shadow\fR.
The user may select alternate files with the \fIpasswd\fR and \fIshadow\fR
parameters.
Additionally, the user may execute the command in read-only mode by
specifying the \fB-r\fR flag.
This causes all questions regarding changes to be answered \fBno\fR
without user intervention.
\fBpwck\fR can also sort entries in \fI/etc/passwd\fR and \fI/etc/shadow\fR
by UID. To run it in sort mode pass it \fB-s\fR flag. No checks are
performed then, it just sorts.
.SH FILES
/etc/passwd \- user account information
.br
/etc/shadow \- encrypted password information
.br
/etc/group \- group information
.SH SEE ALSO
.BR group (5),
.BR passwd (5),
.BR shadow (5),
.BR usermod (8)
.SH DIAGNOSTICS
The \fBpwck\fR command exits with the following values:
.IP 0 5
Success
.IP 1 5
Syntax Error
.IP 2 5
One or more bad password entries
.IP 3 5
Cannot open password files
.IP 4 5
Cannot lock password files
.IP 5 5
Cannot update password files
.SH AUTHOR
Julianne Frances Haugh (jockgrrl@ix.netcom.com)


noflushd/NEWS
* noflushd 2.7.5 (09. May 2005)

  - Fix handling of raid and similar meta devices on 2.6 kernels.

* noflushd 2.7.4 (08. Aug 2004)

  - Fix lousy spindown times on 2.6 kernels.
  - Fix SCSI support on 2.6 kernels.
  - Fix pid detection in generic startup script.
  - Document quirks with hotplugged drives.

* noflushd 2.7.3 (09. Apr 2004)

  - Fix race condition that might leave kupdated/pdflushd disabled
    aftern noflushd has terminated.
  - Fix spindown when all disks are auto-detected.

* noflushd 2.7.2 (16. Mar 2004)

  - A change in 2.7.1 erroneously flushed the buffer cache before
    spindown, causing frequent spinups. This change has been reverted.
    
* noflushd 2.7.1 (07. Mar 2004)

  - Fix immediate spinup on 2.6 kernels.
  - Elevate SCSI support from compile-time to command line option.
  - Bug fixes in the disk statistics parser for 2.6 kernels.
  - Ship kernel patch to track down which processes cause spinups.

* noflushd 2.7 (20. Jan 2004)

  - Support Linux kernel 2.6.x.
  - Default to monitorring all disks unless specific disks are given
    on the command line. This change should yield a sane default
    behaviour for a majority of users without requiring explicit
    configuration.

* noflushd 2.6.3 (13. May 2002)

  - Bugfix for arm and other archs with weird /proc/partitions layout.

* noflushd 2.6.2 (20. Apr 2002)

  - Add IDE raid support.
  - Add timeout logging script.

* noflushd 2.6.1 (03. Feb. 2002)

  - Fix BUG when using noflushd with disks on second IDE channel on 2.4
    kernels.

* noflushd 2.6 (22. Jan 2002)

  - Document noflushd's suckage with any journaling fs, including ext3.
  - New feature prevents spindown if machine is in interactive use.
  - Add hack fix some '0 spindown time' uglies.

* noflushd 2.5 (06. Aug 2001)

  - Make sure hotplugged disks get synced properly.
  - Only sync present partitions. (Fixes high CPU usage bug on devfs
    systems.)
  - Generic startup skript works on latest 2.4 kernels.

* noflushd 2.4 (13. Mar 2001)

  - Timeout lists are back to a working state.
  - Support very latest 2.4 kernels.

* noflushd 2.3 (12. Jan 2001)

  - Support latest 2.4 kernels. (Leaves kernels 2.3.99pre8 to 2.4.0-test8
    unsupported.)
  - Fix multi-disk support with long bdflush timeouts.

* noflushd 2.2 (22. Dec 2000)

  - Fixed bug in spindown code. Now actually spins down. Duh!

* noflushd 2.1 (20. Dec 2000)

  - Disks get synced prior to spindown.
  - Improved devfs support.

* noflushd 2.0 (14. Nov 2000)

  - noflushd now comes with its own syncer that takes over the task from
    kupdate when the first read-write mounted disk is spun down. This
    enhances spindown times in a multidisk environment at the expense of
    less smooth and more bursty background flushing when the noflushd 
    syncer is in use.

  - Conflicting entries in 2.2-style /proc/stat (e.g. /dev/sda and /dev/hda)
    won't cause noflushd to give up anymore. Instead, noflushd will treat
    both disks pretty much like one, i.e. they're either both up or spun
    down, no matter if activity only goes to one of them.

  - Compilation environment switched to automake/autoconf.

  - When installing via 'make install', noflushd is no longer added to
    any runlevel. 

  - noflushd moved to SourceForge. The project homepage is located at
    http://noflushd.sourceforge.net, a mailing list is available via
    noflushd-devel@lists.sourceforge.net. Visit the homepage for 
    subscription information.



shadow-4.0.3/src/pwck.c
/*
 * Copyright 1992 - 1994, Julianne Frances Haugh
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of Julianne F. Haugh nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY JULIE HAUGH AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL JULIE HAUGH OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <config.h>

#include "rcsid.h"
RCSID (PKG_VER "$Id: pwck.c 6 2005-03-20 15:34:28Z bubulle $")
#include <stdio.h>
#include <fcntl.h>
#include <grp.h>
#include "prototypes.h"
#include "defines.h"
#include "chkname.h"
#include <pwd.h>
#include "commonio.h"
#include "pwio.h"
extern void __pw_del_entry (const struct commonio_entry *);
extern struct commonio_entry *__pw_get_head (void);

#ifdef SHADOWPWD
#include "shadowio.h"
extern void __spw_del_entry (const struct commonio_entry *);
extern struct commonio_entry *__spw_get_head (void);
#endif

/*
 * Exit codes
 */

#define	E_OKAY		0
#define	E_USAGE		1
#define	E_BADENTRY	2
#define	E_CANTOPEN	3
#define	E_CANTLOCK	4
#define	E_CANTUPDATE	5

/*
 * Local variables
 */

static char *Prog;
static const char *pwd_file = PASSWD_FILE;

#ifdef	SHADOWPWD
static const char *spw_file = SHADOW_FILE;
#endif
static int read_only = 0;
static int quiet = 0;		/* don't report warnings, only errors */

/* local function prototypes */
static void usage (void);
static int yes_or_no (void);

/*
 * usage - print syntax message and exit
 */

static void usage (void)
{
#ifdef SHADOWPWD
	fprintf (stderr, _("Usage: %s [-q] [-r] [-s] [passwd [shadow]]\n"),
		 Prog);
#else
	fprintf (stderr, _("Usage: %s [-q] [-r] [-s] [passwd]\n"), Prog);
#endif
	exit (E_USAGE);
}

/*
 * yes_or_no - get answer to question from the user
 */

static int yes_or_no (void)
{
	char buf[80];

	/*
	 * In read-only mode all questions are answered "no".
	 */

	if (read_only) {
		puts (_("No"));
		return 0;
	}

	/*
	 * Get a line and see what the first character is.
	 */

	if (fgets (buf, sizeof buf, stdin))
		return buf[0] == 'y' || buf[0] == 'Y';

	return 0;
}

/*
 * pwck - verify password file integrity
 */

int main (int argc, char **argv)
{
	int arg;
	int errors = 0;
	int deleted = 0;
	struct commonio_entry *pfe, *tpfe;
	struct passwd *pwd;
	int sort_mode = 0;

#ifdef	SHADOWPWD
	struct commonio_entry *spe, *tspe;
	struct spwd *spw;
	int is_shadow = 0;
#endif

	/*
	 * Get my name so that I can use it to report errors.
	 */

	Prog = Basename (argv[0]);

	setlocale (LC_ALL, "");
	bindtextdomain (PACKAGE, LOCALEDIR);
	textdomain (PACKAGE);

	OPENLOG (Prog);

	/*
	 * Parse the command line arguments
	 */

	while ((arg = getopt (argc, argv, "eqrs")) != EOF) {
		switch (arg) {
		case 'e':	/* added for Debian shadow-961025-2 compatibility */
		case 'q':
			quiet = 1;
			break;
		case 'r':
			read_only = 1;
			break;
		case 's':
			sort_mode = 1;
			break;
		default:
			usage ();
		}
	}

	if (sort_mode && read_only) {
		fprintf (stderr, _("%s: -s and -r are incompatibile\n"),
			 Prog);
		exit (E_USAGE);
	}

	/*
	 * Make certain we have the right number of arguments
	 */

#ifdef	SHADOWPWD
	if (optind != argc && optind + 1 != argc && optind + 2 != argc)
#else
	if (optind != argc && optind + 1 != argc)
#endif
		usage ();

	/*
	 * If there are two left over filenames, use those as the password
	 * and shadow password filenames.
	 */

	if (optind != argc) {
		pwd_file = argv[optind];
		pw_name (pwd_file);
	}
#ifdef SHADOWPWD
	if (optind + 2 == argc) {
		spw_file = argv[optind + 1];
		spw_name (spw_file);
		is_shadow = 1;
	} else if (optind == argc)
		is_shadow = spw_file_present ();
#endif

	/*
	 * Lock the files if we aren't in "read-only" mode
	 */

	if (!read_only) {
		if (!pw_lock ()) {
			fprintf (stderr, _("%s: cannot lock file %s\n"),
				 Prog, pwd_file);
			if (optind == argc)
				SYSLOG ((LOG_WARN, "cannot lock %s",
					 pwd_file));
			closelog ();
			exit (E_CANTLOCK);
		}
#ifdef	SHADOWPWD
		if (is_shadow && !spw_lock ()) {
			fprintf (stderr, _("%s: cannot lock file %s\n"),
				 Prog, spw_file);
			if (optind == argc)
				SYSLOG ((LOG_WARN, "cannot lock %s",
					 spw_file));
			closelog ();
			exit (E_CANTLOCK);
		}
#endif
	}

	/*
	 * Open the files. Use O_RDONLY if we are in read_only mode, O_RDWR
	 * otherwise.
	 */

	if (!pw_open (read_only ? O_RDONLY : O_RDWR)) {
		fprintf (stderr, _("%s: cannot open file %s\n"),
			 Prog, pwd_file);
		if (optind == argc)
			SYSLOG ((LOG_WARN, "cannot open %s", pwd_file));
		closelog ();
		exit (E_CANTOPEN);
	}
#ifdef	SHADOWPWD
	if (is_shadow && !spw_open (read_only ? O_RDONLY : O_RDWR)) {
		fprintf (stderr, _("%s: cannot open file %s\n"),
			 Prog, spw_file);
		if (optind == argc)
			SYSLOG ((LOG_WARN, "cannot open %s", spw_file));
		closelog ();
		exit (E_CANTOPEN);
	}
#endif

	if (sort_mode) {
		pw_sort ();
#ifdef	SHADOWPWD
		if (is_shadow)
			spw_sort ();
#endif
		goto write_and_bye;
	}

	/*
	 * Loop through the entire password file.
	 */

	for (pfe = __pw_get_head (); pfe; pfe = pfe->next) {
		/*
		 * If this is a NIS line, skip it. You can't "know" what NIS
		 * is going to do without directly asking NIS ...
		 */

		if (pfe->line[0] == '+' || pfe->line[0] == '-')
			continue;

		/*
		 * Start with the entries that are completely corrupt.  They
		 * have no (struct passwd) entry because they couldn't be
		 * parsed properly.
		 */

		if (!pfe->eptr) {

			/*
			 * Tell the user this entire line is bogus and ask
			 * them to delete it.
			 */

			printf (_("invalid password file entry\n"));
			printf (_("delete line `%s'? "), pfe->line);
			errors++;

			/*
			 * prompt the user to delete the entry or not
			 */

			if (!yes_or_no ())
				continue;

			/*
			 * All password file deletions wind up here. This
			 * code removes the current entry from the linked
			 * list. When done, it skips back to the top of the
			 * loop to try out the next list element.
			 */

		      delete_pw:
			SYSLOG ((LOG_INFO, "delete passwd line `%s'",
				 pfe->line));
			deleted++;

			__pw_del_entry (pfe);
			continue;
		}

		/*
		 * Password structure is good, start using it.
		 */

		pwd = pfe->eptr;

		/*
		 * Make sure this entry has a unique name.
		 */

		for (tpfe = __pw_get_head (); tpfe; tpfe = tpfe->next) {
			const struct passwd *ent = tpfe->eptr;

			/*
			 * Don't check this entry
			 */

			if (tpfe == pfe)
				continue;

			/*
			 * Don't check invalid entries.
			 */

			if (!ent)
				continue;

			if (strcmp (pwd->pw_name, ent->pw_name) != 0)
				continue;

			/*
			 * Tell the user this entry is a duplicate of
			 * another and ask them to delete it.
			 */

			puts (_("duplicate password entry\n"));
			printf (_("delete line `%s'? "), pfe->line);
			errors++;

			/*
			 * prompt the user to delete the entry or not
			 */

			if (yes_or_no ())
				goto delete_pw;
		}

		/*
		 * Check for invalid usernames.  --marekm
		 */
		if (!check_user_name (pwd->pw_name)) {
			printf (_("invalid user name '%s'\n"),
				pwd->pw_name);
			errors++;
		}

		/*
		 * Make sure the primary group exists
		 */

		if (!quiet && !getgrgid (pwd->pw_gid)) {

			/*
			 * No primary group, just give a warning
			 */

			printf (_("user %s: no group %u\n"),
				pwd->pw_name, pwd->pw_gid);
			errors++;
		}

		/*
		 * Make sure the home directory exists
		 */

		if (!quiet && access (pwd->pw_dir, F_OK)) {

			/*
			 * Home directory doesn't exist, give a warning
			 */

			printf (_
				("user %s: directory %s does not exist\n"),
				pwd->pw_name, pwd->pw_dir);
			errors++;
		}

		/*
		 * Make sure the login shell is executable
		 */

		if (!quiet && pwd->pw_shell[0]
		    && access (pwd->pw_shell, F_OK)) {

			/*
			 * Login shell doesn't exist, give a warning
			 */

			printf (_("user %s: program %s does not exist\n"),
				pwd->pw_name, pwd->pw_shell);
			errors++;
		}
	}

#ifdef	SHADOWPWD
	if (!is_shadow)
		goto shadow_done;

	/*
	 * Loop through the entire shadow password file.
	 */

	for (spe = __spw_get_head (); spe; spe = spe->next) {
		/*
		 * If this is a NIS line, skip it. You can't "know" what NIS
		 * is going to do without directly asking NIS ...
		 */

		if (spe->line[0] == '+' || spe->line[0] == '-')
			continue;

		/*
		 * Start with the entries that are completely corrupt. They
		 * have no (struct spwd) entry because they couldn't be
		 * parsed properly.
		 */

		if (!spe->eptr) {

			/*
			 * Tell the user this entire line is bogus and ask
			 * them to delete it.
			 */

			printf (_("invalid shadow password file entry\n"));
			printf (_("delete line `%s'? "), spe->line);
			errors++;

			/*
			 * prompt the user to delete the entry or not
			 */

			if (!yes_or_no ())
				continue;

			/*
			 * All shadow file deletions wind up here. This code
			 * removes the current entry from the linked list.
			 * When done, it skips back to the top of the loop
			 * to try out the next list element.
			 */

		      delete_spw:
			SYSLOG ((LOG_INFO, "delete shadow line `%s'",
				 spe->line));
			deleted++;

			__spw_del_entry (spe);
			continue;
		}

		/*
		 * Shadow password structure is good, start using it.
		 */

		spw = spe->eptr;

		/*
		 * Make sure this entry has a unique name.
		 */

		for (tspe = __spw_get_head (); tspe; tspe = tspe->next) {
			const struct spwd *ent = tspe->eptr;

			/*
			 * Don't check this entry
			 */

			if (tspe == spe)
				continue;

			/*
			 * Don't check invalid entries.
			 */

			if (!ent)
				continue;

			if (strcmp (spw->sp_namp, ent->sp_namp) != 0)
				continue;

			/*
			 * Tell the user this entry is a duplicate of
			 * another and ask them to delete it.
			 */

			puts (_("duplicate shadow password entry\n"));
			printf (_("delete line `%s'? "), spe->line);
			errors++;

			/*
			 * prompt the user to delete the entry or not
			 */

			if (yes_or_no ())
				goto delete_spw;
		}

		/*
		 * Make sure this entry exists in the /etc/passwd
		 * file.
		 */

		if (!pw_locate (spw->sp_namp)) {

			/*
			 * Tell the user this entry has no matching
			 * /etc/passwd entry and ask them to delete it.
			 */

			puts (_("no matching password file entry\n"));
			printf (_("delete line `%s'? "), spe->line);
			errors++;

			/*
			 * prompt the user to delete the entry or not
			 */

			if (yes_or_no ())
				goto delete_spw;
		}

		/*
		 * Warn if last password change in the future.  --marekm
		 */

		if (!quiet && spw->sp_lstchg > time ((time_t *) 0) / SCALE) {
			printf (_
				("user %s: last password change in the future\n"),
				spw->sp_namp);
			errors++;
		}
	}

      shadow_done:
#endif

	/*
	 * All done. If there were no deletions we can just abandon any
	 * changes to the files.
	 */

	if (deleted) {
	      write_and_bye:
		if (!pw_close ()) {
			fprintf (stderr, _("%s: cannot update file %s\n"),
				 Prog, pwd_file);
			SYSLOG ((LOG_WARN, "cannot update %s", pwd_file));
			closelog ();
			exit (E_CANTUPDATE);
		}
#ifdef	SHADOWPWD
		if (is_shadow && !spw_close ()) {
			fprintf (stderr, _("%s: cannot update file %s\n"),
				 Prog, spw_file);
			SYSLOG ((LOG_WARN, "cannot update %s", spw_file));
			closelog ();
			exit (E_CANTUPDATE);
		}
#endif
	}

	/*
	 * Don't be anti-social - unlock the files when you're done.
	 */

#ifdef	SHADOWPWD
	if (is_shadow)
		spw_unlock ();
#endif
	(void) pw_unlock ();

	/*
	 * Tell the user what we did and exit.
	 */

	if (errors)
#ifdef NDBM
		printf (deleted ?
			_
			("%s: the files have been updated; run mkpasswd\n")
			: _("%s: no changes\n"), Prog);
#else
		printf (deleted ?
			_("%s: the files have been updated\n") :
			_("%s: no changes\n"), Prog);
#endif

	closelog ();
	exit (errors ? E_BADENTRY : E_OKAY);
}
